Muehle_Algo.ipynb

```diff
@@ -208,33 +278,59 @@
-    "* $\\alpha \\leq \\texttt{value_ab}(state, player) \\leq \\beta \\;\\rightarrow\\;\\texttt{alpha_beta}(state, player, \\alpha, \\beta) = \\texttt{value}(state,player)$\n",
-    "* $\\texttt{value}(state, player) < \\alpha \\;\\rightarrow\\; \\texttt{alpha_beta}(state, player, \\alpha, \\beta) \\leq \\alpha$\n",
+    "Die Funktion `alpha_beta(state, player, alpha, beta)` erhält die gleichen Werte wie die Funktion `value_ab()` und gibt den Wert für den übergebenen Zustand zurück. Dabei arbeitet `alpha_beta()` nach den im Skript vorgestellten Regeln:\r\n",
+    "* $\\alpha \\leq \\texttt{value}(state, player) \\leq \\beta \\;\\rightarrow\\;\\texttt{alpha_beta}(state, player, \\alpha, \\beta) = \\texttt{value}(state,player)$\r\n",
+    "* $\\texttt{value}(state, player) < \\alpha \\;\\rightarrow\\; \\texttt{alpha_beta}(state, player, \\alpha, \\beta) \\leq \\alpha$\r\n",
     "* $\\beta < \\texttt{value}(state, player) \\;\\rightarrow\\; \\beta \\leq \\texttt{alpha_beta}(state, player, \\alpha, \\beta)$"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 10,
    "metadata": {},
    "outputs": [],
    "source": [
-    "def alpha_beta(state, player, alpha, beta, depth):\n",
-    "    state = to_list(state)\n",
-    "    if finished(state, player):\n",
-    "        return utility(state, player)\n",
-    "    if depth == 0:\n",
-    "        return heuristic(state, player)\n",
-    "    value = alpha\n",
-    "    for ns in next_states(state, player):\n",
-    "        value = max(value, -value_ab(ns, opponent(player), -beta, -alpha, depth-1))\n",
-    "        if value >= beta:\n",
-    "            return value\n",
-    "        alpha = max(value, alpha)\n",
+    "def alpha_beta(state, player, alpha, beta, depth):\r\n",
+    "    state = to_list(state)\r\n",
+    "    if finished(state, player):\r\n",
+    "        return utility(state, player)\r\n",
+    "    if depth == 0:\r\n",
+    "        return heuristic(state, player)\r\n",
+    "    value = alpha\r\n",
+    "\r\n",
+    "    #move-ordering\r\n",
+    "    nextStates = []\r\n",
+    "    for s in next_states(state, player):\r\n",
+    "        nextStates = nextStates + [[s, get_order_value(s, player, depth)]]\r\n",
+    "    nextStates = sorted(nextStates,key=itemgetter(1),reverse=True)\r\n",
+    "\r\n",
+    "    for ns in nextStates:\r\n",
+    "        value = max(value, -value_ab(ns[0], opponent(player), -beta, -alpha, depth-1))\r\n",
+    "        if value >= beta:\r\n",
+    "            return value\r\n",
+    "        alpha = max(value, alpha)\r\n",
     "    return value"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Die Funktion `get_order_value(state, player, depth)` erhält den Spielzustand, einen Spieler und die Suchtiefe, und gibt den Wert für diesen Zustand aus der vorherigen Iteration zurück."
```

Comment:
Die Funktion wird in alpha-beta verwendet, aber erst später definiert. Bitte nach oben ziehen